### ISSUE TITLE
option to install from source

### ISSUE THREAD

**Issue by orodbhen** [conda/conda#1328]
It would be useful to provide a conda install option that recompiles the package(s) before installing. Similar to what pip does.

Expecting a single binary package to run on any platform is a little overly ambitious in some cases. This might work OK for Python packages, but for conda to be a means of installing any package into a  semi self-contained environment, I think it really needs to have this capability. 

It seems like this would be failry simple to implement. Just download the package, extract the recipe, and go. Unfortunately, in the case of many Anaconda packages, this won't work because the recipes are in the wrong format. Moreover, it's pretty wasteful to do it this way, and it would be better to only download the recipe files before running conda build.

**Comment by asmeurer** [comment]
So you're saying that conda should have an option to build a recipe automatically. But if someone has a recipe that works for a given platform, couldn't they have just built it and uploaded it to binstar anyway?

**Comment by seibert** [comment]
There can be non-trivial issues with glibc that make it hard to distribute truly portable packages.  I've been in situations where I couldn't easily deal with the range of Linux distributions that my colleagues were using.

**Comment by orodbhen** [comment]
Like @selbert said, building binaries that will run on any hardware running the same version of the same Linux distribution is ambitious enough. But trying to create a binary that will work on any operating system is, in some cases, nearly impossible. 

For example, the gcc package in the Anaconda repo fails to compile a large number of packages, at least in part because its features.h file is outdated. Probably because it was originally compiled on an older Linux OS (CentOS 5, I think). My understanding is that the output of GCC compilation, including header files, depends a lot on where it's compiled. 

Recompiling the package manually isn't too hard, if it's only the one package, but often you need to recompile the dependencies as well.

I realize that downloading the binary packages and recompiling them would be a waste of bandwidth. I'm thinking you could have something  like the Arch Linux AUR repo, where the packages only contain recipes. It would be a separate channel.

**Comment by stuarteberg** [comment]
@orodbhen:

>For example, the gcc package in the Anaconda repo fails to compile a large number of packages, at least in part because its features.h file is outdated.

If you are interested in getting it to work on your system, see conda/conda-recipes#279

**Comment by orodbhen** [comment]
@stuarteberg, thanks, that looks promising. I only skimmed the thread, but I'll go through in detail later.

I actually attempted to create a combined build of the whole toolchain using conda build. Strangely, it fails to build with conda, but compiles fine running "make, make install" manually, with PREFIX=/usr.

Since the combined build pulls in all the dependencies from source, I'm thinking that the problem is caused by changes that conda makes to the build environment. Anyway, I should submit a separate issue for this.
